lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,772 @@
+  /// the instructions between the end_borrow and the destroy_value are users of
+  /// %original_borrowee.  That requires having a set of uses to check for
+  /// membership in, though, so that is postponed until
+  /// isViableInstructionSequence.
+  ///
+  /// returns the destroy_value instruction in the sequence that destroys
+  ///         %original_borrowee or nullptr if there isn't one
+  DestroyValueInst *
+  mayBeginOriginalBorroweeDestroyingSequence(SILInstruction *) const;
+
+  /// Slow check, dependent on finding users, that a Match found by
+  /// definesMatchingInstructionSequence can be folded.
+  MatchViability isViableInstructionSequence(
+      Match &, SmallVectorImpl<int> &rewritableArgumentIndices) const;
+
+  /// Find among the scope ending instructions of %lifetime any that match the
+  /// expected instruction sequence pattern.
+  ///
+  /// Uses definesMatchingInstructionSequence to determine whether an
+  /// instruction is a candidate for folding.
+  ///
+  /// Includes both full matches which could potentially be folded and partial
+  /// matches which need to be verified later aren't illegal in
+  /// isViableInstructionSequence.
+  ///
+  /// returns true if any full matches are found
+  ///         false otherwise
+  bool findCandidates(SmallVectorImpl<Candidate> &candidates) const;
+
+  /// Identifies all the simple extended users of %introducer.
+  ///
+  /// Populates: users, extendedSimpleUseCountPerApplySite.
+  ///
+  /// returns true if none of the uses of %introducer escaped
+  ///         false otherwise
+  bool findUsers();
+
+  /// Determines whether each candidate is viable for folding.
+  ///
+  /// Requiring findUsers to have been called, this uses the more expensive
+  /// isViableInstructionSequence.
+  ///
+  /// returns true if any candidates were viable
+  ///         false otherwise
+  bool filterCandidates(SmallVectorImpl<Candidate> &candidates) const;
+
+  /// Find all uses of %original_borrowee that are dominated by introducer.
+  ///
+  /// We are only interested in those dominated by introducer because those are
+  /// the uses all of which must be "outside" the liveness boundary of
+  /// %introducer.  In detail, PrunedLiveness::isWithinBoundary relies on
+  /// clients to know that instructions are after the start of liveness.  We
+  /// determine this via the dominance tree.
+  ///
+  /// Populates: originalBorroweeUses.
+  bool findDominatedUsesOfOriginalBorrowee();
+
+  /// Whether there are any uses of the original borrowee within the borrow
+  /// scope.
+  ///
+  /// If there are, we can't fold the apply.  Specifically, we can't introduce
+  /// a move_value %original_borrowee because that value still needs to be used
+  /// in those locations.  While updateSSA would happily rewrite those uses to
+  /// be uses of the move, that is not a semantically valid transformation.
+  bool originalBorroweeHasUsesWithinBorrowScope();
```

Comment:
I've already forgotten the SIL pattern that breaks if you don't check originalBorroweeHasUsesWithinBorrowScope. I *think* it had to do with another overlapping borrow scope, but don't quite see how that's possible. Can you elaborate just a bit in the comment?

Reply:
Added.  It's just any use at all of the borrowee in the live range of the borrow.